<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">

    <meta property=”og:title” content="Portfolio || Bioinformatics B-tree"/>
    <meta property=”og:type” content=”website” />
    <meta property="og:image" content="https://portingdata.com/image_files/open_graph_image.png"/>
    <meta property="og:url" content="https://portingdata.com/ipi"/>

    <meta name="description" content="Portfolio || Predict Future Sales">

    <title>Porting Data</title>
    <link href="style.css" type="text/css" rel="stylesheet">
    <link href="icons.css" type="text/css" rel="stylesheet">
    <link href="textfield.css" type="text/css" rel="stylesheet">
    <link href="predict_sales.css" type="text/css" rel="stylesheet">
</head>
<body>
    <!-- Header Strip -->
    <div class="header">
        <div class="title">
            <h1>Porting Data  | |  </h1><h1 id="header_title">Bioinformatics B-tree</h1>
        </div>
        <div></div>
        <div>
            <!-- Navigation Bar -->
            <nav class="navbar">            
                <a href="index.html">Home</a>
                <a href="ipi_report.html">Strategic Insight</a>
                <a href="predict_sales.html">Predict Sales</a>
                
            </nav>
        </div>
    </div>


    
    <section class="body">
        <!-- Icon Sidebar -->
        <div id="icons">
            <div class="row">
              <h4>Dominik Huffield</h4>
            </div> 
            <a href="https://github.com/the-chiefdom" target="_blank"> 
            <img class="logostyle_a" src="image_files/github_logo.png"><p>Github</p>
            </a>
            <a href="https://www.linkedin.com/in/dominik-huffield" target="_blank">
            <img class="logostyle_a" src="image_files/linkedin_logo.png"><p>LinkedIn</p>
            </a>
            <a href="https://github.com/the-chiefdom/Acoustics_FTB" target="_blank">
            <img class="logostyle_b" src="image_files/tensoraudio_logo.png"><p>Tensor Audio</p>
            </a>
            <a href="image_files/Dominik_Resume_2021.pdf" target="_blank">
            <img class="logostyle_a" src="image_files/resume_logo.png"><p>Resume</p>
            </a>
        </div>

        <div id="textfield_pfs" class="textfield">
            <!-- Financial Spending Model -->
            <h2 id="h2_small">Bioinformatics</h2>
            
            <section id="sec_small" class="h2_sec">
                <p>
                    This project reads the Human Genome by encoding the sequence. 
                    Sequenced using the 4 different organic chemicals, known as bases, the Human Genome is about 2.87 billion bases long. 
                    Utilizing a B-tree data structure as a solution for the memory constraints by writing to disk we will also be able to search the sequence with a O(log n) capability.
                </p>     
                <p>View the <a href="https://github.com/the-chiefdom/Bioinformatics" target="_blank" style="color: #3366cc;">Project Repository</a> for a deeper dive into the code</p>
                
                <!-- Purpose, motivation and description: -->
                <h3>Background</h3>
                <section class="h3_sec">
                    <p> 
                        Bioinformatics is the field of science in which biology, computer science, and information technology merge to form a single discipline. 
                        One of the primary aims of Bioinformatics is to attempt to determine the structure and meaning of the human genome. 
                        The human genome is a complete set of human DNA. 
                        The Human Genome project was started in 1990 by the United States Department of Energy and the U.S. National Institutes of Health. 
                        By April 14, 2003 99% of the Human Genome had been sequenced with 99.9% accuracy.
                        <br><br>
                        The Human Genome is a big strand of 4 different organic chemicals, known as bases, which are:
                        <br>
                        Adenine, Cytosine, Thiamine, Guanine
                        <br><br>
                        Biologists often call them A, C, T, G for short. The bases A and T are always paired together. 
                        Similarly the bases C and G are always paired together. 
                        So when we look at the DNA representation, only one side is listed. 
                        For example: the DNA sequence: AATGC actually represents two sequences: AATGC and its complement TTACG (replace A by T, T by A, C by G and G by C). 
                        Even with only half the bases represented in a DNA sequence, the human genome is about 2.87 billion characters long! 
                        See Figure 1 for a image of the DNA as well as the chemical structure of the bases.
                        <br><br>
                        The primary source for getting the human genome (as well as all other mapped organisms) is in the National Center for Biotechnology Information (NCBI) website.
                        <br>
                        <a href="http://www.ncbi.nlm.nih.gov/" target="_blank" style="color: #3366cc;">http://www.ncbi.nlm.nih.gov/</a>
                        <br><br>
                        We will be using the GeneBank files from NCBI. The format is described with a sample file at the following web link:
                        <br>
                        <a href="http://www.ncbi.nlm.nih.gov/Sitemap/samplerecord.html" target="_blank" style="color: #3366cc;">http://www.ncbi.nlm.nih.gov/Sitemap/samplerecord.html</a>
                        
                    </p>
                    <img class="text_img" src="image_files/bioinfo.png">
                    
                </section>

                <!-- Input Files: -->
                <h3>Input Files</h3>
                <section class="h3_sec">
                    <p>
                        The GeneBank files have a bunch of annotations followed by the keyword ORIGIN. 
                        The DNA sequences start from the next line. 
                        Each line has 60 characters (one of A, T, C, G, could be lower/upper case) until the end of sequence, which is denoted by // on a line by itself. 
                        Sometimes you will see the character N, which denotes that the sequence is not known at that character. 
                        You would skip these characters. 
                        One GeneBank file may have several DNA sequences in it, each marked by ORIGIN and // tags.
                        <br><br>
                        When we reach a character N, we assume that the subsequence has ended. 
                        Similarly, when we reach //, we also assume that the subsequence has ended. 
                        So at those points, we reset the subsequence that we were building and start over when we find the next ORIGIN or the next valid character after seeing a N.
                    </p>
                </section>

                <!-- Problem: -->
                <h3>Problem</h3>
                <section class="h3_sec">
                    <p>
                        For a given GeneBank file, we want to convert it into a BTree with each object being a DNA sequence of specified length k (where 1 ≤ k ≤ 31). 
                        We will take the DNA sequence from the GeneBank file and break it into sequences of length k each. 
                        We are interested in all subsequences with length k. 
                        For example, in the sequence AATTCG, the subsequences of length three are: AAT, ATT, TTC and TCG. 
                        Once we have a BTree for a length k, we want to be able to search for query sequences of length k. 
                        The search returns the frequency of occurrence of the query string (which can be zero if it is not found).
                        <br><br>
                        The biological motivation behind is to study the frequency of different length subsequences to see if they are random or that some sequences are more likely to be found in the DNA.
                    </p>
                </section>
                
                <!-- Design Issues: -->
                <h3>Design Issues</h3>
                <section class="h3_sec">
                    <p>
                        Saving memory. Since we only have four possible bases (A, C, G and T), we can optimize on space by converting our DNA strings to base 4 numbers. 
                        Then we can represent each base by a 2-bit binary number.
                        <br><br>
                        Note that we have made the binary representations for complementary bases be binary complements as well. 
                        With this compact representation, we can store a 31 length sequence in a 64-bit long primitive type in Java.
                        <br><br>
                        Key Values. Note that the binary compact representation of the subsequences will result in a unique 64-bit integer value. 
                        Hence we can directly use that as our key value.
                        <br><br>
                        Class Design. We will need a BTree class as well as a BTreeNode class. 
                        The objects that we store in the BTree will be similar to the objects we stored in the previous Hashtable assignment. 
                        You may call the relevant class TreeObject to represent the objects.
                    </p>
                </section>
                
                <!-- Implementation: -->
                <h3>Implementation</h3>
                <section class="h3_sec">
                    <p>
                        There are two seperate programs: one that creates a BTree from a given GeneBank file and another for searching in the specified BTree for sequences of given length. 
                        The search program assumes that the user specified the proper BTree to use depending upon the query length.
                        <br><br>
                        There are two main Java classes named GeneBankCreateBTree and GeneBankSearch.
                        <br>
                        The required arguments for the two programs are shown below.
                        <br><br>
                        java GeneBankCreateBTree  &lt;0/1(no/with Cache)&gt; &lt;degree&gt; &lt;gbk file&gt; &lt;sequence length&gt;
                            [&lt;cache size&gt;] [&lt;debug level&gt;]
                        <br><br>
                        java GeneBankSearch &lt;0/1(no/with Cache)&gt; &lt;btree file&gt; &lt;query file&gt; [&lt;cache size&gt;]
                            [&lt;debug level&gt;]
                        <br<br>
                        The degree is the degree to be used for the BTree. 
                        If the user specifies 0, then your program should choose the optimum degree based on a disk block size of 4096 bytes and the size of your BTree node on disk.
                        <br><br>
                        The sequence length is an integer that must be between 1 and 31 (inclusive). 
                        The query file contains all the DNA strings of a specific sequence length that we want to search for in the specified btree file. 
                        The strings are one per line and they all must have the same length as the DNA sequences in the btree file. 
                        The DNA strings use a, c, t, and g (either lower or upper case).
                        <br><br>
                        The debug level is an optional argument with a default value of zero. 
                        It must at least support the following levels for GeneBankCreateBTree:
                    </p>
                </section> 
            </section>   
            <p></p>
        </div> 
        
    </section>

    <footer>
        <p>&copy; 2020 Dominik Huffield</p>
    </footer>
</body>
</html>